src/mock/mod.rs

```diff
@@ -0,0 +1,323 @@
+//! smaller than 2^31), but should not produce accidental collisions at an unacceptable rate.
+//!
+//! As a result, all "cryptographic" operations can be carried out much faster. This module is
+//! intended to be used during unit-tests of applications that build on top of `threshold_crypto`;
+//! enabling this in production code of any application will immediately break its cryptographic
+//! security.
+
+pub mod ms8;
+
+use std::{fmt, mem, slice};
+
+use pairing::{EncodedPoint, GroupDecodingError};
+use pairing::{Field, PrimeField};
+use rand;
+
+use super::{CurveAffine, CurveProjective, Engine};
+
+pub use self::ms8::Mersenne8;
+
+/// A `pairing` Engine based on `Mersenne8` prime fields.
+#[derive(Clone, Debug)]
+pub struct Mocktography;
+
+/// Affine type for `Mersenne8`.
+#[derive(Copy, Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Ms8Affine(Mersenne8);
+
+/// Projective type for `Mersenne8`.
+#[derive(Copy, Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Ms8Projective(Mersenne8);
+
+impl fmt::Display for Ms8Affine {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self.0, f)
+    }
+}
+
+impl fmt::Display for Ms8Projective {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(&self.0, f)
+    }
+}
+
+impl rand::Rand for Ms8Affine {
+    #[inline]
+    fn rand<R: rand::Rng>(rng: &mut R) -> Self {
+        Ms8Affine(rng.gen())
+    }
+}
+
+impl rand::Rand for Ms8Projective {
+    #[inline]
+    fn rand<R: rand::Rng>(rng: &mut R) -> Self {
+        Ms8Projective(rng.gen())
+    }
+}
+
+impl From<Ms8Projective> for Ms8Affine {
+    fn from(Ms8Projective(x): Ms8Projective) -> Ms8Affine {
+        Ms8Affine(x)
+    }
+}
+
+impl From<Ms8Affine> for Ms8Projective {
+    fn from(Ms8Affine(x): Ms8Affine) -> Ms8Projective {
+        Ms8Projective(x)
+    }
+}
+
+impl Engine for Mocktography {
+    type G1 = Ms8Projective;
+    type G1Affine = Ms8Affine;
+    type G2 = Ms8Projective;
+    type G2Affine = Ms8Affine;
+    type Fq = Mersenne8;
+    type Fqe = Mersenne8;
+    type Fqk = Mersenne8;
+
+    // In newer versions of pairing, this must be moved to `ScalarEngine`:
+    type Fr = Mersenne8;
+
+    fn miller_loop<'a, I>(_i: I) -> Self::Fqk
+    where
+        I: IntoIterator<
+            Item = &'a (
+                &'a <Self::G1Affine as CurveAffine>::Prepared,
+                &'a <Self::G2Affine as CurveAffine>::Prepared,
+            ),
+        >,
+    {
+        // Unused?
+        unimplemented!()
+    }
+
+    fn final_exponentiation(_: &Self::Fqk) -> Option<Self::Fqk> {
+        // Unused?
+        unimplemented!()
+    }
+
+    fn pairing<G1, G2>(p: G1, q: G2) -> Self::Fqk
+    where
+        G1: Into<Self::G1Affine>,
+        G2: Into<Self::G2Affine>,
+    {
+        p.into().0 * q.into().0
+    }
+}
+
+impl AsRef<[u64]> for Mersenne8 {
```

Comment:
Should these `impl AsRef`s and `impl Mut`s be in `src/mock/ms8.rs` alongside `Mersenne8`s other impls?

Reply:
Ah, that was my bad. You're probably right.

Reply:
Ideally, we would not implement these at all, they are an affordance for implementing the traits from `pairing`. All the `AsRef`s are a bit tied to how `pairing` implements these things for their own structures.